packages/dds/tree/src/feature-libraries/forestRepairDataStore.ts

```diff
@@ -0,0 +1,191 @@
+                        }
+                        case Delta.MarkType.MoveOut:
+                        case Delta.MarkType.Delete: {
+                            onDelete(parent, key, index, mark.count);
+                            index += mark.count;
+                            break;
+                        }
+                        case Delta.MarkType.Modify: {
+                            cursor.enterNode(index);
+                            const child = parent.getOrCreateChild(key, index, undefinedFactory);
+                            visitModify(mark, child);
+                            cursor.exitNode();
+                            index += 1;
+                            break;
+                        }
+                        case Delta.MarkType.Insert:
+                        case Delta.MarkType.InsertAndModify:
+                        case Delta.MarkType.MoveIn:
+                        case Delta.MarkType.MoveInAndModify:
+                            break;
+                        default:
+                            unreachableCase(type);
+                    }
+                }
+            }
+        }
+
+        function visitModify(modify: ModifyLike, node: RepairDataNode): void {
+            // Note that the `in` operator return true for properties that are present on the object even if they
+            // are set to `undefined. This is leveraged here to represent the fact that the value should be set to
+            // `undefined` as opposed to leaving the value untouched.
+            if ("setValue" in modify) {
+                if (node.data === undefined) {
+                    node.data = repairDataFactory();
+                }
+                const value = cursor.value;
+                if (node.data.value === undefined) {
+                    node.data.value = new Map();
+                }
+                node.data.value.set(revision, value);
+            }
+            if (modify.fields !== undefined) {
+                visitFieldMarks(modify.fields, node);
+            }
+        }
+
+        function onDelete(
+            parent: RepairDataNode,
+            key: FieldKey,
+            startIndex: number,
+            count: number,
+        ): void {
+            for (let i = 0; i < count; ++i) {
+                const fork = cursor.fork();
+                const index = startIndex + i;
+                fork.enterNode(index);
+                const nodeData = mapTreeFromCursor(fork);
+                fork.free();
+                const child = parent.getOrCreateChild(key, index, repairDataFactory);
+                if (child.data === undefined) {
+                    child.data = repairDataFactory();
+                }
+                if (child.data.node === undefined) {
+                    child.data.node = new Map();
+                }
+                child.data.node.set(revision, nodeData);
+            }
+        }
+
+        visitFieldMarks(destruction.changes, this.root);
```

Comment:
I was planning on doing that, but it ended up being a realy pain. The awkward thing about using the visitor is that it give you indices assuming you've applied any deletes and inserts at lower indices, which is not what we want here since we're not mutating the forest. So to use the visitor you'd have to subtract the impact of those prior marks from the index being passed to the visitor callback. Since the visitor bounces up and down the tree it means you have to maintain a data stucture to associate such a compensating offset with each field you have visited.

Reply:
I see. Maybe we should see if we can add a feature to DeltaVisitor to allow it to better support this scenario.

Reply:
I suppose we could pass to the visitor both indices (the currently passed one which assumes the edits have been carried out, and the input context index which is what is needed here). Another negative effect of using the visitor is that it forces this code to be more stateful (doing push/pops to simulate what the callstack is doing here). We could extend the visitor contract to help with that too (the "enter" methods would return something that the "exit" method would be passed back). Do you think that's worth doing?